nullaway/src/main/java/com/uber/nullaway/handlers/FieldInitializationSerializationHandler.java

```diff
@@ -0,0 +1,83 @@
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+package com.uber.nullaway.handlers;
+
+import com.google.errorprone.VisitorState;
+import com.google.errorprone.util.ASTHelpers;
+import com.sun.source.tree.Tree;
+import com.sun.source.util.TreePath;
+import com.sun.tools.javac.code.Symbol;
+import com.uber.nullaway.NullabilityUtil;
+import com.uber.nullaway.dataflow.AccessPathNullnessAnalysis;
+import com.uber.nullaway.fixserialization.FixSerializationConfig;
+import com.uber.nullaway.fixserialization.out.FieldInitializationInfo;
+import javax.lang.model.element.ElementKind;
+
+/**
+ * This handler is used to serialize information regarding methods that initialize a class field. If
+ * a method guarantee to leave the initialized class field in the method body to be {@code @NonNull}
+ * at exit point, this handler will serialize information regarding the initializer method and the
+ * class field. These information helps to detect initializer methods in classes.
+ */
+public class FieldInitializationSerializationHandler extends BaseNoOpHandler {
+
+  private final FixSerializationConfig config;
+
+  FieldInitializationSerializationHandler(FixSerializationConfig config) {
+    this.config = config;
+  }
+
+  /**
+   * If the method guarantees to leave the initialized class field to be {@code @Nonnull} at exit
+   * point, this method will serialize information regarding the initializer method and the class
+   * field. Since traversing AST is costly, we do it only inside the handler when the feature is
+   * enabled and, this method accesses the initializer method through the leaf node in state
+   * parameter.
+   */
+  @Override
+  public void onNonNullFieldAssignment(
+      Symbol field, AccessPathNullnessAnalysis analysis, VisitorState state) {
+    TreePath pathToMethod =
+        NullabilityUtil.findEnclosingMethodOrLambdaOrInitializer(state.getPath());
+    if (pathToMethod == null || pathToMethod.getLeaf().getKind() != Tree.Kind.METHOD) {
+      return;
+    }
+    Symbol.MethodSymbol symbol = (Symbol.MethodSymbol) ASTHelpers.getSymbol(pathToMethod.getLeaf());
+    // We are only looking for non-constructor methods that initializes a class field.
+    if (symbol.getKind() == ElementKind.CONSTRUCTOR) {
+      return;
+    }
+    final String fieldName = field.getSimpleName().toString();
+    boolean leavesNonNullAtExitPoint =
```

Comment:
Before this: it's worth checking that the field is owned by same `ClassSymbol` as the method containing the assignment. We don't want to allow `A.m(...)` to be a candidate initializer for `B.f` unless `A == B`, correct? Regardless of whether, say, `A` is an inner class of `B` or a factory for `B` (An argument can be made for subclasses of abstract classes, but NullAway won't acknowledge `@Initializer` annotations in that case anyways).

Edit: Note that checking for the same `getSimpleName()` is not enough, there could be a `B.f` and `A.f` and both be distinct fields.

Reply:
7e6570acf37e44ab29f29c4aada0e963631a0245

Not sure, if calling `findEnclosingNode` is the best way to check this issue since it may decrease the performance. Please let me know if there is a better way.

Reply:
Thank you very much. Looks much better now, sorry I forgot about `Symbol.enclClass()`.

b02bb563c90b4caf22c86bc15a58f252bd8d6dc2